### PR TITLE
KAFKA-16763: Upgrade to scala 2.12.19 and scala 2.13.14

### PR DESCRIPTION
**Pull request by m1a2st**
Scala has been releas new version, so upgrade scala version

### CODE REVIEW

**Comment by m1a2st** [comment]
@chia7712, This PR please take a look, Thank you.

**Comment by chia7712** [comment]
```
[2024-05-15T01:51:49.255Z] > Task :core:compileScala

[2024-05-15T01:51:49.255Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/controller/KafkaController.scala:1202:18: method setOrCreatePartitionReassignment in class KafkaZkClient is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:518:14: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:524:24: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:533:23: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:533:49: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated

```

please file the build error

**Comment by m1a2st** [comment]
@chia7712, Thanks for your review, Scala 2.3 build is passed in Jenkins

File: build.gradle [hunk 1]
```diff
@@ -671,7 +671,8 @@ subprojects {
       "-Xlint:private-shadow",
       "-Xlint:stars-align",
       "-Xlint:type-parameter-shadow",
-      "-Xlint:unused"
+      "-Xlint:unused",
```
Comment (chia7712) [inline build.gradle]: Pardon me, why we need those changes?

Comment (m1a2st) [reply build.gradle]: I catch two main errors are class is depreate (1. ) and `@nowarn` build error  in previous build in Jenkins, but these classes or method are used by other classes. I consider that I should not migrate these methods or classes in this PR, thus I think that should closed these check when building scala version 2.3.
1.  ([Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/controller/KafkaController.scala:1202:18: method setOrCreatePartitionReassignment in class KafkaZkClient is deprecated)

Comment (chia7712) [reply build.gradle]: The issue is caused by the magic `NotNothing`. As we are in modern scala, could you try to remove it? I give a try on my local, and both compile and test work well.

Comment (m1a2st) [reply build.gradle]: Ok, I will remove the `"-Wconf:cat=unused-nowarn:s"`, and try to complie

Comment (m1a2st) [reply build.gradle]:
```
[2024-05-16T08:37:59.531Z] [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/network/RequestChannel.scala:174:69: @nowarn annotation does not suppress any warnings

[2024-05-16T08:37:59.531Z] one error found
```

This error happened on Jenkins.

Comment (chia7712) [reply build.gradle]: @m1a2st You have to remove the usage of `NotNothing`

**Review by chia7712 — APPROVED** [review]
LGTM

File: core/src/main/scala/kafka/network/RequestChannel.scala [hunk 2]
```diff
@@ -171,7 +170,7 @@ object RequestChannel extends Logging {
       s"$forwardDescription$header -- ${loggableRequest.toString(details)}"
     }
 
-    def body[T <: AbstractRequest](implicit classTag: ClassTag[T], @nowarn("cat=unused") nn: NotNothing[T]): T = {
+    def body[T <: AbstractRequest](implicit classTag: ClassTag[T]): T = {
```
Comment (ijuma) [inline core/src/main/scala/kafka/network/RequestChannel.scala:173]: Why did we remove this? It was there for a reason (to prevent cases where `ClassTag` is automatically inferred to `Nothing`). Same for all other similar cases.

File: build.gradle [hunk 3]
```diff
@@ -651,7 +651,7 @@ subprojects {
     scalaCompileOptions.keepAliveMode = userKeepAliveMode
 
     scalaCompileOptions.additionalParameters = [
-      "-deprecation",
+      "-deprecation:false",
```
Comment (ijuma) [inline build.gradle:654]: Why was this changed?

Comment (m1a2st) [reply build.gradle:654]: In new scala version, it will check the deprecated method, and make the compile error like below
```
[2024-05-15T01:51:49.255Z] > Task :core:compileScala

[2024-05-15T01:51:49.255Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/controller/KafkaController.scala:1202:18: method setOrCreatePartitionReassignment in class KafkaZkClient is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:518:14: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:524:24: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:533:23: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated

[2024-05-15T01:51:50.590Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/zk/ZkData.scala:533:49: class LegacyPartitionAssignment in object ReassignPartitionsZNode is deprecated
```
To resolve this, I found the way is to  add `"-deprecation:false"`

File: core/src/main/scala/kafka/network/RequestChannel.scala (hunk 2 above)
Comment (m1a2st) [reply core/src/main/scala/kafka/network/RequestChannel.scala:173]: In new Scala version, it will check `@nowarn` annotation, and make that build error. 
```
[2024-05-16T08:37:59.531Z] [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-15958/core/src/main/scala/kafka/network/RequestChannel.scala:174:69: @nowarn annotation does not suppress any warnings
[2024-05-16T08:37:59.531Z] one error found
```
In previous comments, we have two way to resolve this.
1. "-Wconf:cat=unused-nowarn:s" 
2. Remove the usage of NotNothing

we decide to use the second way to resolve this problem

File: build.gradle (hunk 3 above)
Comment (chia7712) [reply build.gradle:654]: Another way is to enable that check, and then we fix all related warnings (most of fixes are to add `nowarn`)

File: core/src/main/scala/kafka/network/RequestChannel.scala (hunk 2 above)
Comment (chia7712) [reply core/src/main/scala/kafka/network/RequestChannel.scala:173]: > It was there for a reason (to prevent cases where ClassTag is automatically inferred to Nothing)

That was a good protection, but we use the request once casting the `body`. Hence, the `nothing` type (i.e request.body`) will be noticed soon as we can't call any methods from it. 

> we decide to use the second way to resolve this problem

As @m1a2st mentioned, there is another approach for it. We can go back to use it if @ijuma prefer it
